### PR TITLE
feat(acir)!: Verify Proof

### LINKED ISSUE #198: Add recursion opcode
### Problem

The ability to verify a program in another program is one feature which modern programming languages have. As with all opcodes, this can be done in the frontend using the arithmetic opcode, however it will be in-efficient and there are ways in which the proving system can do it in a more optimal way. One of these is delaying the pairing computation until verification, so it is not computed inside of the circuit.

### Proposed solution

**An alternative**

Lets first discuss an alternative which may eventually deprecate the strategy we will go with now.

The idea is to look at what parts of the verification algorithms we _really_ need to delegate to the proving system. In the case that the proving system uses [KZG](https://link.springer.com/chapter/10.1007/978-3-642-17373-8_11) we would only create a pairings opcode and the frontend will efficiently compute the rest. An example of a frontend would be Noir. It is the case that there are other components in the verification algorithm that are expensive like big integer arithmetic and this solution would hinge on us having enough opcodes such that they cover the possible usecases such as big integer.

The reason to not go with this strategy _right_ now is because we would need to do a survey of the known proving systems and figure out what components we really need to delegate to the proving system.

The advantages of this strategy is that you more of the code is implemented in the frontend or in the acvm-stdlib, making it less opaque.

**Proposed solution** 

To implement a `VERIFY_PROOF` opcode which will take a proof, public inputs and a verification key. It will then return true to indicate whether the proof is successful and false to indicate whether the proof is not successful.

**Structures**

This will be a new black box function, where the function definition will be:

```rust
FuncDefinition {
           name : "verify_proof",
           input_size: InputSize::Variable,
           output_size: OutputSize(1),
},
```

The `input_size` will be the concatenation of the verification key, proof and public inputs. The input size is variable for two reasons:

- The public inputs can vary depending on the circuit.
- (Not ideal) we don't specify what proving system we are verifying a proof for, so this recursion opcode can be used to verify plonk/groth16/bulletproofs for example. This is not ideal because if a frontend changes the backend but does not change the proof, then they will encounter a failing proof.

For the latter point, we could introduce an opcode for each proof system that we can recurse on though this is not an ideal solution because each time we add a backend which is supported, we would then need to add an opcode.

### Alternatives considered

_No response_

### Additional context

_No response_

### Submission Checklist

- [X] Once I hit submit, I will assign this issue to the Project Board with the appropriate tags.

### PR DESCRIPTION
**Pull request by vezenovm**
# Related issue(s)

Resolves #198 

The `acvm-backend-barretenberg` PR which implements this can be found here:
https://github.com/noir-lang/acvm-backend-barretenberg/pull/207
The relevant Noir PR which implements this and the backend PR can be found here:
https://github.com/noir-lang/noir/pull/1491

# Description

Add VerifyProof variant to BlackBoxFunc enum

## Summary of changes

(Describe the changes in this PR. Point out breaking changes if any.)

## Dependency additions / changes

(If applicable.)

## Test additions / changes

(If applicable.)

# Checklist

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [ ] I have reviewed the changes on GitHub, line by line.
- [ ] I have ensured all changes are covered in the description.

# Additional context

(If applicable.)

### CODE REVIEW

File: acir/src/circuit/opcodes/black_box_function_call.rs [hunk 1]
```diff
@@ -192,6 +211,32 @@ impl BlackBoxFuncCall {
...
+                        + 1
+                        + input_aggregation_object.len(),
+                );
+                inputs.extend(key.iter().copied());
+                inputs.extend(proof.iter().copied());
+                inputs.extend(public_inputs.iter().copied());
+                inputs.push(*key_hash);
+                // If we do not have an input aggregation object assigned
+                // do not return it as part of the input vector as to not trigger
+                if !input_aggregation_object.iter().any(|v| v.witness == Witness(0)) {
```
Comment (kevaundray) [inline acir/src/circuit/opcodes/black_box_function_call.rs:235]: Witness(0) seems to be a barretenberg specific idiom?

File: acvm/src/lib.rs [hunk 2]
```diff
@@ -150,6 +159,7 @@ pub trait ProofSystemCompiler {
         circuit: &Circuit,
         witness_values: WitnessMap,
         proving_key: &[u8],
+        is_recursive: bool,
```
Comment (kevaundray) [inline acvm/src/lib.rs:162]: Why is this needed and can it not be deduced from the circuit?

File: acvm/src/lib.rs [hunk 3]
```diff
@@ -160,5 +170,18 @@ pub trait ProofSystemCompiler {
         public_inputs: WitnessMap,
         circuit: &Circuit,
         verification_key: &[u8],
+        is_recursive: bool,
     ) -> Result<bool, Self::Error>;
+
+    fn proof_as_fields(
```
Comment (kevaundray) [inline acvm/src/lib.rs:176]: Why is this and the below function needed?

Comment (vezenovm) [reply acvm/src/lib.rs:176]: We need a proof and verification key to be witnesses of the circuit in order to recursively verify them. Most proofs and verification key's are not formatted in a proof system's native field elements. I was debating to put this in `acvm-backend-barretenberg` but it felt universal enough for being able to perform recursion with Noir in general. 

@zac-williamson do you think these methods are too specific for the acvm and should possibly just be moved into our backend? For example, if a halo2 backend were to implement recursion do you think these methods would still be necessary?

File: acvm/src/lib.rs (hunk 2 above)
Comment (vezenovm) [reply acvm/src/lib.rs:162]: Unfortunately it cannot really be deduced from the circuit. An inner circuit may require a different prover in order to be verified recursively (in the case of bberg the inner circuit prover must be based upon Pedersen-Blake3 rather than Keccak). The internal circuit is able to just be a normal circuit that looks no different from any other Noir circuit

File: acir/src/circuit/opcodes/black_box_function_call.rs (hunk 1 above)
Comment (vezenovm) [reply acir/src/circuit/opcodes/black_box_function_call.rs:235]: So I was debating how to go about this here. I could add a flag that determines whether or not we should include the `input_aggregation_object` in our inputs vector instead of this check. If the `input_aggregation_object` is all zeros we will get a missing assignment error when we call `solve` in `blackbox.rs`. Setting the `input_aggregation_object` as all zeros is specific to our bberg recursion circuit, but either way we are going to have to some sort of indicator whether to have an input aggregation object assigned. I was thinking a flag is similar to requiring anyone who is integrating the VerifyProof blackbox func to have the `input_aggregation_object` of their first recursive verification be all zeros. However, a flag is more explicit and I can switch to this

File: acvm/src/lib.rs (hunk 2 above)
Comment (kevaundray) [reply acvm/src/lib.rs:162]: what does is_recursive denote/mean?

File: acvm/src/lib.rs (hunk 3 above)
Comment (kevaundray) [reply acvm/src/lib.rs:176]: How/Where would these two methods be called?

File: acir/src/circuit/opcodes/black_box_function_call.rs (hunk 1 above)
Comment (kevaundray) [reply acir/src/circuit/opcodes/black_box_function_call.rs:235]: We don't have anything in acvm right now that states that Witness(0) is special

Can we have `input_aggregation_object` be optional then explain what it means for it to be None?

File: acir/src/circuit/opcodes/black_box_function_call.rs [hunk 4]
```diff
@@ -76,6 +76,16 @@ pub enum BlackBoxFuncCall {
         inputs: Vec<FunctionInput>,
         outputs: Vec<Witness>,
     },
+    VerifyProof {
+        key: Vec<FunctionInput>,
```
Comment (kevaundray) [inline acir/src/circuit/opcodes/black_box_function_call.rs:80]:
```suggestion
        verification_key: Vec<FunctionInput>,
```

File: acir/src/circuit/opcodes/black_box_function_call.rs [hunk 5]
```diff
@@ -76,6 +76,16 @@ pub enum BlackBoxFuncCall {
...
+        proof: Vec<FunctionInput>,
+        public_inputs: Vec<FunctionInput>,
```
Comment (kevaundray) [inline acir/src/circuit/opcodes/black_box_function_call.rs:82]: Add a comment here specifying what these `public_inputs` refer to

File: acir/src/circuit/opcodes/black_box_function_call.rs [hunk 6]
```diff
@@ -76,6 +76,16 @@ pub enum BlackBoxFuncCall {
...
+        key_hash: FunctionInput,
```
Comment (kevaundray) [inline acir/src/circuit/opcodes/black_box_function_call.rs:83]: Add comment specifying whether this is the proving/verification key hash and also why it is needed

File: acir/src/circuit/opcodes/black_box_function_call.rs [hunk 7]
```diff
@@ -76,6 +76,16 @@ pub enum BlackBoxFuncCall {
         inputs: Vec<FunctionInput>,
         outputs: Vec<Witness>,
     },
+    VerifyProof {
```
Comment (kevaundray) [inline acir/src/circuit/opcodes/black_box_function_call.rs:79]: I don't think VerifyProof is the right name for this opcode, since that would indicate that this should return a boolean as output and or at the least that once this opcode was called, that the proof was verified

File: acir/src/circuit/opcodes/black_box_function_call.rs [hunk 8]
```diff
@@ -76,6 +76,16 @@ pub enum BlackBoxFuncCall {
...
+        key: Vec<FunctionInput>,
+        proof: Vec<FunctionInput>,
+        public_inputs: Vec<FunctionInput>,
+        key_hash: FunctionInput,
+        input_aggregation_object: Vec<FunctionInput>,
+        // This is the recursive verification output aggregation object.
+        // The name `outputs` was kept to simplify code reuse with the other BlackBoxFuncCall's
```
Comment (kevaundray) [inline acir/src/circuit/opcodes/black_box_function_call.rs:86]: I think changing this to output_aggregation_object would be better especially since we no longer have a homogenous API for the black box functions

File: acir/src/circuit/opcodes/black_box_function_call.rs (hunk 7 above)
Comment (vezenovm) [reply acir/src/circuit/opcodes/black_box_function_call.rs:79]: Hmm yes good point. We could call it `VerifyRecursiveProof` or `RecursiveVerification` (although I do `VerifyProof` implies this as well)? I will leave comments indicating the purpose of the input and output aggregation object

File: acir/src/circuit/opcodes/black_box_function_call.rs (hunk 8 above)
Comment (vezenovm) [reply acir/src/circuit/opcodes/black_box_function_call.rs:86]: I will change it to the specific name

File: acir/src/circuit/opcodes/black_box_function_call.rs (hunk 7 above)
Comment (kevaundray) [reply acir/src/circuit/opcodes/black_box_function_call.rs:79]: Hmm it seems to be aggregating and not entirely verifying

**Review by kevaundray — COMMENTED** [review]
Generally looks okay -- Some overall questions:

- Can you give a brief outline as to how this opcode will be used from the user perspective? (I'm trying to understand the rationale behind this interface)

- The size of the aggregation objects can change depending on the proving system, where do we account for this?

File: acvm/src/lib.rs (hunk 3 above)
Comment (vezenovm) [reply acvm/src/lib.rs:176]: These would be called by someone looking to pass a proof and/or verification key into their Noir circuit. The way our recursive verification opcode is setup we require a verification key and a proof and these must be represented by witness values. 

Any backend is going to have its own recursion format for a proof and/or vkeys. Being able to export the proof and verification key in the appropriate format will be necessary for any backend looking to integrate recursion. I could move this functionality to acvm-backend-bberg but then I will still want to expose functionality for a backend to be able to generate the appropriate recursive format for a proof and vkey. 

Having these methods be a part of the ACVM allows for us to more easily integrate it into `nargo`. Otherwise this will be a purely backend specific functionality, when it really is not. 

Let's say I have circuit A. I compile and generate a proof for this circuit.

I want to verify circuit A in circuit B. I need to pass in the proof and vkey in the appropriate recursion format. These methods will be used to generate the proof and vkey as fields (perhaps I should rename the functions `***_in_recursion_format`?). The outputted proof and vkey as fields will be input to circuit B and used as parameters for the verify proof builtin.

File: acvm/src/lib.rs (hunk 2 above)
Comment (vezenovm) [reply acvm/src/lib.rs:162]: It means we want to generate a circuit which will be recursively verified in another circuit. But how it is being used right now is essentially to determine the hash type of the prover for an inner circuit. I do agree though this is not general enough and I will consider if there is a better way to abstract this functionality. What are your thoughts on a `hash_type` flag that determines the fiat shamir hash used by the prover?

File: acir/src/circuit/opcodes/black_box_function_call.rs (hunk 7 above)
Comment (vezenovm) [reply acir/src/circuit/opcodes/black_box_function_call.rs:79]: We do verify the previous aggregation state technically, and then aggregate the current proof onto it. So aggregation may be a better name. `RecursiveAggregation`?

File: acir/src/circuit/opcodes/black_box_function_call.rs (hunk 1 above)
Comment (vezenovm) [reply acir/src/circuit/opcodes/black_box_function_call.rs:235]: Yes I can switch to that instead

File: acvm/src/lib.rs (hunk 2 above)
Comment (vezenovm) [reply acvm/src/lib.rs:162]: So the inner proof that is generated may use a different algorithm than a normal proof that can be verified by the Solidity verifier. The inner proof in this case may use a different hash function for Fiat-Shamir as that hash is more snark friendly and more optimized for recursive verification. 

This is information that would be passed to the proof system, however, there is no way to indicate from the circuit alone that we want to use a different hash type for the prover. I think if the `is_recursive` flag is too specific (and I agree is a bit of a leakage from bberg) we will need some other flag. This flag can either be a `hash_type` flag as proposed above or a custom data index of some sort. 

Perhaps a custom_data index of type [u8] would be simplest and then we just leave it for whichever backend is implementing these methods to indicate what custom data means. I imagine we would then want the backend to expose some sort of custom_data API as nargo is going to have to integrate these methods

Comment (kevaundray) [reply acvm/src/lib.rs:162]: Its mainly because when you create proofs that you want to be natively verfied, you use perhaps sha256 for your hash function in the proof system. But when you want to verify the proof in a snark, you want the hash function being used to be snark friendly for the fiat-shamir/transcript. is_recursive will use a snark-friendly hash function when true and otherwise it will use the regular one, which for bb is keccak, but it can be anything for a proving system

File: acir/src/circuit/opcodes/black_box_function_call.rs (hunk 1 above)
Comment (vezenovm) [reply acir/src/circuit/opcodes/black_box_function_call.rs]: I switched to an Option. It is now entirely up to a backend to determine how they will handle the first recursive aggregation

File: acvm/src/lib.rs (hunk 3 above)
Comment (vezenovm) [reply acvm/src/lib.rs:178]: I added a comment on this function. It might not be descriptive enough so let me know what you think

File: acir/src/circuit/opcodes/black_box_function_call.rs (hunk 4 above)
Comment (vezenovm) [reply acir/src/circuit/opcodes/black_box_function_call.rs]: changed

File: acir/src/circuit/opcodes/black_box_function_call.rs (hunk 5 above)
Comment (vezenovm) [reply acir/src/circuit/opcodes/black_box_function_call.rs:85]: Comments added

File: acir/src/circuit/opcodes/black_box_function_call.rs (hunk 6 above)
Comment (vezenovm) [reply acir/src/circuit/opcodes/black_box_function_call.rs:89]: Commented

File: acir/src/circuit/opcodes/black_box_function_call.rs (hunk 7 above)
Comment (vezenovm) [reply acir/src/circuit/opcodes/black_box_function_call.rs]: I changed the name to `RecursiveAggregation`

File: acir/src/circuit/opcodes/black_box_function_call.rs (hunk 8 above)
Comment (vezenovm) [reply acir/src/circuit/opcodes/black_box_function_call.rs]: Changed

File: acir/src/circuit/black_box_functions.rs [hunk 9]
```diff
@@ -44,6 +45,7 @@ impl BlackBoxFunc {
             BlackBoxFunc::XOR => "xor",
             BlackBoxFunc::RANGE => "range",
             BlackBoxFunc::Keccak256 => "keccak256",
+            BlackBoxFunc::VerifyProof => "verify_proof",
```
Comment (vezenovm) [inline acir/src/circuit/black_box_functions.rs:48]: I changed the name of the `BlackBoxFuncCall` to `RecursiveAggregation` but I was unsure on changing the `BlackBoxFunc` call to that as well. Leaving the builtin with the name `verify_proof` I feel like would still be good but then the mismatch with `BlackBoxFuncCall::RecursiveAggregation` is a bit deceiving.

File: acvm/src/lib.rs (hunk 2 above)
Comment (vezenovm) [reply acvm/src/lib.rs:166]: I added some comments to explain `is_recursive`

File: acvm/src/lib.rs [hunk 10]
```diff
@@ -101,6 +101,15 @@ pub trait PartialWitnessGenerator {
         input: &FunctionInput,
         outputs: &[Witness],
     ) -> Result<OpcodeResolution, OpcodeResolutionError>;
+    fn verify_proof(
```
Comment (kevaundray) [inline acvm/src/lib.rs:104]: I forget the conclusion of our discussion re having this done inside of the backend. 

I'll information dump for context.

Removing it gives an imbalance because the RecursiveAggregation opcode will be the only one without a solver method.

Keeping it means that the solver will either directly or indirectly need a backend (unless the solver implements it to return true) + it will be a component that will now need to be setted when we attempt to solve.

The root root problem being that there is no standard way to do recursive aggregation, so its not possible to implement it here like the other opcodes. However, since these witnesses are never used by the business logic in the program (fed into another opcode) we can optimistically assume that it verifies true and ignore it.

The conversation over slack was that we needed to set the output values `output_aggregation_object`using this method, but it was noted that we do not need to set these until we get to the backend since they are never used in other places in the IR, so we could remove this from the ACVM layer altogether.

Comment (vezenovm) [reply acvm/src/lib.rs:104]: I think we are safe to remove this from the ACVM layer. However, it does feel weird for `RecursiveAggregation` to not have a solver method. I can remove this from the ACVM for now and leave a comment describing why it is unnecessary and our choice to always have the recursion opcode pass execution during simulation.

File: acir/src/circuit/black_box_functions.rs (hunk 9 above)
Comment (kevaundray) [reply acir/src/circuit/black_box_functions.rs:48]: Yeah its somewhat confusing I think, especially since its not verifying a proof

**Review by kevaundray — APPROVED** [review]
